### ISSUE TITLE
Access Denied when running in AspNetCore as an IIS website

### ISSUE THREAD

**Issue by dubeg** [mgaffigan/Rmg.PdfPrinting#2]
Hi Mitch, I know you never intended this project to run in an aspnetcore website under IIS, but I still tried it & figured I should post the results. 

When trying to print in these conditions, an Access Denied error occurs, & I'm not exactly sure how to explain it. Is it that an IIS process cannot create/close the D2D print control?

```
System.UnauthorizedAccessException: Access is denied. (0x80070005 (E_ACCESSDENIED))
     at Windows.Win32.Graphics.Direct2D.ID2D1PrintControl.Close()
     at Rmg.PdfPrinting.PdfPrinter.RenderDocToPrintControl(PdfDocument pdfDoc, ID2D1PrintControl printControl)
     at Rmg.PdfPrinting.PdfPrinter.Print(String printerName, PdfDocument pdfDoc, String jobName, Stream ticket)
```

On the other hand, printing works fine when running the aspnetcore app under my user account (for eg. when debugging in Visual Studio).

**Comment by mgaffigan** [comment]
Hi @dubeg,

Thanks for opening an issue. Direct2D is [definitely supported](https://learn.microsoft.com/en-us/windows/win32/direct2d/direct2d-portal) from IIS, but I'm not sure about the WinRT PDF APIs. 

I'd start by checking the basics: can `IIS AppPool\Whatever` open the PDF file and any destination files?

You might try using procmon to compare what happens when run in IIS vs IIS Express.  Feel free to attach captures here if you don't see anything.

I'm on the go this week, so I doubt I'll be able to setup an IIS server to test.

**Comment by dubeg** [comment]
Hey Mitch, thanks for the response. I just tried using Load Full Profile, no luck there. The pdf I’m trying to print is in a stream (memory stream, I think, but I’ll have to verify that) which I then  generated using a WebView2 control nagivating to an URL and converting the html into a pdf. I’ll post a sample project tomorrow if you’d like. 

How could I demystify that Access Denied otherwise? I don't ETL tracing but could I potentially use that to determine the source of the error? Or perhaps process monitor.

```

```

**Comment by dubeg** [comment]
Ha, I've found the cause.

Googling around, the conditions triggering the issue seem to be:
- Running the app in 32-bit 
- Running under a non-standard user account
- Accessing the CPrintTicket WoW Services (AppID: `{AA0B85DA-FDDF-4272-8D1D-FF9B966D75B0}`).

In my case, I fixed it by adding permissions on that COM server for my IIS AppPool identity via the Component Services GUI.
It's been an issue since 2013, I don't think this will ever get a fix from Microsoft:P

References:

- https://mskb.pkisolutions.com/kb/2872151
- https://stackoverflow.com/questions/20390495/web-browser-print-from-service-running-on-windows-2012
- https://stackoverflow.com/questions/15834064/printing-from-windows-service-on-windows-8-fails
